teps/0050-ignore-task-failures.md

```diff
@@ -187,6 +194,222 @@ control over the task definitions but may desire to ignore a failure and continu
 
   ![Jenkins Dashboard](images/0050-jenkins-dashboard-with-failure-stage.png)
 
+## Proposal
+We propose a new filed ```OnFail``` to the [PipelineTask](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#adding-tasks-to-the-pipeline) definition.
+
+```go
+type PipelineTask struct {
+	Name string `json:"name,omitempty"`
+
+	//...
+
+	// OnFail defines the termination behavior of a pipeline when the task is failed
+	// can be set to [ continue | stopAndFail ]
+	OnFail OnFailType `json:onFail, "omitempty"`
+}
+```
+
+```go
+type OnFailType string
+
+const (
+	// StopAndFail indicates to stop the pipeline if the task is failed
+	StopAndFail OnFailType = "stopAndFail"
+	// Continue indicates to continue executing the rest of the pipeline irrespective of the status of the task
+	Continue    OnFailType = "continue"
+)
+```
+
+Pipeline author can set the ```OnFailType``` field to configure the task failure strategy. If set to ```StopAndFail```, the pipeline is stopped and failed when the task is failed. If set to ```Continue```, the failure of task is ignored and the pipeline continues to execute the rest of the DAG.
+
+```yaml
+- name: task1
+  OnFail: continue
+  taskSpec:
+    steps:
+      - image: alpine
+        name: exit-with-1
+        script: |
+          exit 1
+
+```
+This new field ```OnFail``` will be implemented as a alpha feature and can be enabled by setting enable-api-fields to alpha.
+
+Setting ```OnFail``` is optional, the default pipeline behaviour is ```StopAndFail```
+
+The task run information is avaialble under the ```pipelineRun.status.taskRuns```. Note that the original task run status remains as it is irrelevant of the value of ```OnFail``` (i.e. a failed task with ```OnFail: continue``` is still marked as failed). A prefix "[Task is failed but ignored]" will be added as a prefix to the original ```message``` of the task.
+
```

Comment:
We are utilizing `message` more and more to distinguish the status. For the clients/applications/users consuming taskrun status, its not easy to compare statement vs pre-defined constant.  I would request to introduce a new reason rather than adding such elaborated message. 

If we continue going this route, we will have introduced a bunch of new messages after a couple of years. 

Reference: https://github.com/tektoncd/pipeline/blob/main/docs/taskruns.md#monitoring-execution-status

`status`|`reason`|`message`|`completionTime` is set|Description
:-------|:-------|:--|:---------------------:|--------------:
False|TaskRunCancelled|n/a|Yes|The TaskRun was cancelled successfully.
False|TaskRunCancelled|TaskRun cancelled as the PipelineRun it belongs to has timed out.|Yes|The TaskRun was cancelled because the PipelineRun timed out.
False|TaskRunTimeout|n/a|Yes|The TaskRun timed out.

Reply:
I think this is a good point. However, we should put the real cause of the failure in the ```reason``` field like ```TaskRunTimeout```. Creating a new reason like ```FailureAndIgnored``` does not give the real cause and sounds confusing to users (with ```status:False```). 

Instead, I can create a new string field like ```FailureIgnored``` to [TaskRunStatusFields](https://github.com/tektoncd/pipeline/blob/11e5c376d62e673350d62fab036ea97990a06dba/pkg/apis/pipeline/v1/taskrun_types.go#L197) to indicate if the task run failure is ignored or not. We can also benefit the flexibility from such string field if we want to support more failure strategies in the future (other than ```continue``` and ```stopAndFail```)

Reply:
> create a new string field like FailureIgnored to [TaskRunStatusFields](https://github.com/tektoncd/pipeline/blob/11e5c376d62e673350d62fab036ea97990a06dba/pkg/apis/pipeline/v1/taskrun_types.go#L197)

thanks @QuanZhang-William, wouldn't this be confusing? What we are looking for is to identify the execution status of a `taskRun`.  And `type TaskRunStatus struct {...` is our standard so far to hold the status. In addition to the existing `reasons`:

```
TaskRunReasonFailed TaskRunReason = "Failed"
TaskRunReasonImagePullFailed TaskRunReason = "TaskRunImagePullFailed"
```

We introduce a new reason - `FailureIgnored` :

```
TaskRunReasonFailed TaskRunReason = "FailureIgnored"
```

We can certainly populate message with an additional details if needed.